### PR TITLE
[fsdp, perf] fix PrefixGrouper integration

### PR DESCRIPTION
**Pull request by supercharleszhu**
### What does this PR do?

Fix the [PrefixGrouper](https://github.com/verl-project/verl/blob/352f76f59ff014c6b423eed1a21a7881be1a0b46/examples/prefix_grouper/README.md?plain=1#L22) integration after the worker-to-engine refactor. https://github.com/verl-project/verl/pull/6067

This PR:

1. propagates `use_prefix_grouper` into FSDP/FSDP2 actor and reference engines;
2. builds grouped prompt/response inputs at the current engine boundary;
3. splits grouped hidden states before the vocabulary projection;
4. computes response log-probabilities and entropy without materializing full
  grouped-sequence vocabulary logits;
5. calls the separately wrapped LM-head module so FSDP2 unshard, reshard, and
  backward hooks remain active;
6. reconstructs response outputs into the existing jagged sequence layout before
  the unchanged PPO loss path;
7. propagates the tokenizer pad ID so padded prompt tokens are excluded correctly.

### Context: Engine Refactor causing PrefixGrouper deactivated

The original integration from #4368 branched inside
[`DataParallelPPOActor._forward_micro_batch`](https://github.com/verl-project/verl/blob/6eeb5711babe178ca3cc3caba7c9f158efa9fcd6/verl/workers/actor/dp_actor.py#L129-L148):

```python
if self.use_prefix_grouper:
    can_use_pg = (
        not self.use_remove_padding
        and not self.use_ulysses_sp
        and not self.use_fused_kernels
        and not self.use_dynamic_bsz
    )
    if can_use_pg and "response_mask" in micro_batch and "uid" in micro_batch:
        return forward_micro_batch_with_prefix_grouper(...)
```

That path grouped adjacent responses with the same `uid`, selected one prompt
per group, concatenated each shared prompt with its responses, and passed the
`prefix_grouper` object through the patched attention implementation. The
unchanged legacy
[`pg_forward`](https://github.com/verl-project/verl/blob/6eeb5711babe178ca3cc3caba7c9f158efa9fcd6/verl/trainer/ppo/prefix_grouper_utils.py#L104-L147)
then split the grouped full-vocabulary logits into response tensors and computed
log-probabilities and entropy.

Commit
[`044bbba2`](https://github.com/verl-project/verl/commit/044bbba27e5861e1212727aba67813cccb2db291)
(#6067) deleted `dp_actor.py` and migrated actor/reference execution to the
engine abstraction. The replacement
[`FSDPEngineWithLMHead.prepare_model_inputs`](https://github.com/verl-project/verl/blob/044bbba27e5861e1212727aba67813cccb2db291/verl/workers/engine/fsdp/transformer_impl.py#L885-L965)
implemented remove-padding and fused-kernel paths but did not consume
`use_prefix_grouper` or call the existing helper.

The configuration field and UID-aware trainer balancing remained, so the option
became a silent no-op: repeated responses stayed grouped in the batch, but the
engine still forwarded every repeated prompt independently.

### Context : response-only LM-head projection to save memory

The original PrefixGrouper helper grouped the transformer input correctly, but
projected every grouped prefix, response, and padding position through the
vocabulary-sized LM head before splitting response outputs:

```text
grouped hidden
  -> full [grouped_tokens, vocab_size] logits
  -> split full-vocabulary response logits
  -> convert to FP32
  -> compute target-token log-probabilities and entropy
```

This might easily cause OOM every model. In an 8xH100 Qwen3.5-0.8B test
(`vocab_size=151936`), our production workload has grouped logits shape = [16, 13,592, 151,936] and converting them to FP32 can
require approximately 16* 13592 * 151936 * 4 ~= 123Gb , before accounting for autograd state and other
model tensors.

This PR instead splits grouped transformer hidden states before the LM head and
uses `FusedLinearForPPO` only on response prediction positions:

```text
grouped hidden
  -> split [last prefix hidden + response hidden]
  -> chunked target-token log-probability and entropy projection
  -> no full-vocabulary response-logit tensor
```

On that workload, the completed response-only path will only take response hidden: [64, 256, 2,048]. FusedLinearForPPO  processes 512-token chunks which shape [512, 151,936], so the peak memory will be 296 MB for FP32.

The implementation calls `self.lm_head(...)` rather than directly reading
`self.lm_head.weight`. FSDP2 may wrap the LM head separately, and invoking the
module preserves its parameter unshard, reshard, and backward hooks.

----
This does not duplicate an open contribution:

- PR search: https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+PrefixGrouper
- Issue search: https://github.com/verl-project/verl/issues?q=is%3Aissue+is%3Aopen+PrefixGrouper
- #6401 concerns multi-turn prefix-tree attention; this PR repairs the existing
  fixed-group PrefixGrouper path.

AI assistance was used for investigation, implementation, testing, and review.

### Test

Focused CPU tests:

```bash
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1 PYTHONPATH=. \
  python -m pytest -q tests/trainer/ppo/test_prefix_grouper_utils.py
```

Result: `5 passed`.

Changed-file checks:

```bash
ruff check \
  tests/trainer/ppo/test_prefix_grouper_utils.py \
  verl/models/transformers/monkey_patch.py \
  verl/trainer/ppo/prefix_grouper_utils.py \
  verl/trainer/ppo/ray_trainer.py \
  verl/workers/config/actor.py \
  verl/workers/config/engine.py \
  verl/workers/engine/fsdp/transformer_impl.py \
  verl/workers/engine_workers.py

ruff format --check <same files>
```

Result: passed.

e2e validation with our production workload:

- Qwen3-1.7B, 8xH200, four-step experiment (with our internal dataset):

| Metric | Existing path | PrefixGrouper | Difference |
| --- | ---: | ---: | ---: |
| Actor update | 24.15s | 13.79s | -42.9% |
| Old log-prob | 7.93s | 4.76s | -40.0% |
| Ref log-prob | 8.29s | 4.91s | -40.7% |
| Step time | 89.21s | 71.85s | -19.5% |
| Throughput | 5,156 tok/s | 6,401 tok/s | +24.1% |
| Peak allocation | 92.09 GB | 42.67 GB | -53.7% |

OSS example validation:

- Started from
  [`examples/prefix_grouper/run_qwen3_8b_fsdp.sh`](https://github.com/supercharleszhu/verl/blob/main/examples/prefix_grouper/run_qwen3_8b_fsdp.sh).
- Used the GSM8K train/test parquet files and a mounted Qwen3.5-0.8B checkpoint
  because Qwen3-8B was not available on the test pod.
- Replaced the example's dynamic log-prob/actor batching with fixed
  microbatches divisible by `rollout.n`, matching the current PrefixGrouper
  constraints.
- Ran through the latest V1 TransferQueue trainer for one epoch: all 7/7 steps
  completed successfully.
- Final step: 116.25s, 2,780 tok/s, 8.29 GiB peak actor allocation, and
  15.36 GiB peak reservation.

This run exposed and now covers the V1 boundary where TransferQueue provides
jagged prompt, response, and response-mask tensors. PrefixGrouper converts those
fields to padded tensors for grouping and receives the tokenizer pad ID through
the V1 old-log-prob, reference, and actor-update paths.

### API and Usage Example

```yaml
actor_rollout_ref:
  model:
    use_remove_padding: false
    use_fused_kernels: false
  actor:
    strategy: fsdp2
    use_prefix_grouper: true
    use_dynamic_bsz: false
    shuffle: false
    ppo_micro_batch_size_per_gpu: 64
  ref:
    log_prob_use_dynamic_bsz: false
    log_prob_micro_batch_size_per_gpu: 64
  rollout:
    log_prob_use_dynamic_bsz: false
    log_prob_micro_batch_size_per_gpu: 64
```

Fixed microbatch sizes must be divisible by `rollout.n`.

Current constraints:

- text-only models;
- FSDP/FSDP2;
- fixed microbatches;
- `use_dynamic_bsz=false`;
- `shuffle=false`;
- `use_remove_padding=false`;
- `use_fused_kernels=false`;
- sequence parallel size `1`;
- no `calculate_sum_pi_squared` or top-K distillation.

### Design & Code Changes

```text
UID-aware batch balancing
  -> fixed FSDP microbatch
  -> group repeated prompts and responses
  -> shared-prefix attention
  -> split grouped hidden states
  -> response-only LM-head projection
  -> reconstruct original jagged sequence outputs
  -> existing PPO and policy-loss implementations
```

The advantage estimator, reward calculation, optimizer, and policy-loss
selection are unchanged.

### Checklist Before Starting

- [x] Search for similar PRs.
- [x] Format the PR title as `[{modules}] {type}: {description}`.

### Checklist Before Submitting

- [x] Read the contributing guide.
- [ ] Run the full pre-commit hook suite. Direct Ruff checks were run instead;
  unreviewed third-party hooks were not executed in the enterprise environment.
- [ ] Add or update documentation. No documentation file is changed by this PR.
- [x] Add focused unit tests for the changed execution path.
- [ ] Request CI in the `ci-request` channel when ready.
- [x] Recipe submodule update is not applicable.

### CODE REVIEW

**Comment by wuxibin89** [comment]
@supercharleszhu PrefixGrouper has some limitation, we're exploring MAGI attention https://github.com/verl-project/verl/pull/6689

**Comment by supercharleszhu** [comment]
Thanks @wuxibin89 for the review! Will try it out. Since that PR is in review, Do you recommend we hang this pr and wait for that pr to be merged? Any missing item for Magi attention? Lemme know if there is something I can contribute to.

**Comment by wuxibin89** [comment]
@supercharleszhu Prefer to close this PR and discuss in https://github.com/verl-project/verl/pull/6689
